### ISSUE TITLE
Investigate copy_to_directory performance

### ISSUE THREAD

**Issue by kormide** [aspect-build/bazel-lib#258]
See https://bazelbuild.slack.com/archives/CA31HN1T3/p1665178148461539.

A copy_to_directory on 2500 files is taking on the order of 10s. Investigate what's causing the slowness and see if there are any optimizations that can be made.

**Comment by jwnx** [comment]
I'm experiencing the same problem. Copying 513 files takes consistently ~2s.

<img width="1917" alt="screen 2022-10-11 at 13 10 01" src="https://user-images.githubusercontent.com/4687817/195144070-1fd7ee3f-917d-4c3f-874c-8517d8c81669.png">

**Comment by matthewjh** [comment]
I am also having problems with analysis-phase performance of this rule. Our project is a graph of various ts_project, npm_package_link, npm_package, and js_library targets. When making a change to a BULD file, such as to add a dep, the Bazel process uses 500% CPU for 15-20s.

Profiling shows that most of this time is spent in  copy_to_directory_action via runAnalysisPhase:

**Comment by gregmagolan** [comment]
Should be resolved with [version_matrix](https://github.com/aspect-build/bazel-lib/commit/350408b93d3d4f714f287beebdb4581a7d3ef13e) that is included in the [1.16.0](https://github.com/aspect-build/bazel-lib/releases/tag/v1.16.0) release

**Comment by matthewjh** [comment]
Thanks, @gregmagolan , analysis performance is good now. 🙌

Execution performance is still problematic, though, which is what the OP was about. Are you guys still planning to investigate this? Should we open a new issue or re-open this one?

**Comment by matthewjh** [comment]
It seems to me that, besides simplifying the logic, one thing that would certainly improve perf for larger directories is performing the logic in parallel on different subtrees. At the moment this rule males very poor use of CPU as everything is blocking.  This should be easy to implement (eg in js or go) but I'm not sure how to do it in bash alone.
